File: 2023/day_01/python/day01.py
from aoc import AOC
import math

aoc = AOC(year=2023, day=1)
data = aoc.load()

digits = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
def digit(s):
    return digits.index(s) + 1 if s in digits else s

# part 1

numeric_calibration_values = []
for numbers in data.digits_by_line():
    if not numbers: continue
    numeric_calibration_values.append(numbers[0] * 10 + numbers[-1])

aoc.p1(sum(numeric_calibration_values))

# part 2

calibration_values = []
for line in data.lines():
    first = (-1, math.inf)
    last = (-1, -1)

    for d in digits:
        i = digit(d)
        from_start, from_end = line.find(d), line.rfind(d)
        if from_start >= 0:
            if from_start < first[1]: first = (i, from_start)
            if from_end > last[1]: last = (i, from_end)

        from_start, from_end = line.find(str(i)), line.rfind(str(i))
        if from_start >= 0:
            if from_start < first[1]: first = (i, from_start)
            if from_end > last[1]: last = (i, from_end)
    calibration_values.append(first[0] * 10 + last[0])

aoc.p2(sum(calibration_values))

File: 2023/day_01/swift/Day01.swift
import AdventSupport
import Foundation

public class Year2023Day01: Solver {
	public required init() {}

	// MARK: Part 1

	public func solvePart1(_ input: Input) async throws -> String? {
		input.digitsByLine()
			.reduce(0) { total, digits in
				total + Int("\(digits.first!)\(digits.last!)")!
			}
			.description
	}

	// MARK: Part 2

	public func solvePart2(_ input: Input) async throws -> String? {
		let stringToDigits = [
			"one": "1",
			"two": "2",
			"three": "3",
			"four": "4",
			"five": "5",
			"six": "6",
			"seven": "7",
			"eight": "8",
			"nine": "9",
		]

		let digits: [String] = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"] + Array(stringToDigits.keys)

		return input.lines()
			.reduce(0) { total, line in
				let firstDigit = digits.min {
					(line.range(of: $0)?.lowerBound ?? line.endIndex) < (line.range(of: $1)?.lowerBound ?? line.endIndex)
				}!

				let lastDigit = digits.max {
					(line.ranges(of: $0).last?.upperBound ?? line.startIndex) < (line.ranges(of: $1).last?.upperBound ?? line.startIndex)
				}!

				let calibrationValue = Int("\(stringToDigits[firstDigit] ?? firstDigit)\(stringToDigits[lastDigit] ?? lastDigit)")!

				return total + calibrationValue
			}
			.description
	}
}
